src/main/kotlin/app/revanced/patcher/Context.kt

```diff
@@ -0,0 +1,102 @@
+package app.revanced.patcher
+
+import app.revanced.patcher.apk.Apk
+import app.revanced.patcher.apk.ApkBundle
+import app.revanced.patcher.apk.ResourceFile
+import app.revanced.patcher.util.method.MethodWalker
+import org.jf.dexlib2.iface.Method
+import org.w3c.dom.Document
+import java.io.Closeable
+import java.io.InputStream
+import java.io.StringWriter
+import javax.xml.parsers.DocumentBuilderFactory
+import javax.xml.transform.TransformerFactory
+import javax.xml.transform.dom.DOMSource
+import javax.xml.transform.stream.StreamResult
+
+/**
+ * A common class to constrain [Context] to [BytecodeContext] and [ResourceContext].
+ * @param apkBundle The [ApkBundle] for this context.
+ */
+sealed class Context(val apkBundle: ApkBundle)
+
+/**
+ * A context for the bytecode of an [Apk.Base] file.
+ *
+ * @param apkBundle The [ApkBundle] for this context.
+ */
+class BytecodeContext internal constructor(apkBundle: ApkBundle) : Context(apkBundle) {
+    /**
+     * The list of classes.
+     */
+    val classes = apkBundle.base.bytecodeData.classes
+
+    /**
+     * Create a [MethodWalker] instance for the current [BytecodeContext].
+     *
+     * @param startMethod The method to start at.
+     * @return A [MethodWalker] instance.
+     */
+    fun traceMethodCalls(startMethod: Method) = MethodWalker(this, startMethod)
+}
+
+/**
+ * A context for [Apk] file resources.
+ *
+ * @param apkBundle the [ApkBundle] for this context.
+ */
+class ResourceContext internal constructor(apkBundle: ApkBundle) : Context(apkBundle) {
+
+    /**
+     * Open an [DomFileEditor] for a given DOM file.
+     *
+     * @param inputStream The input stream to read the DOM file from.
+     * @return A [DomFileEditor] instance.
+     */
+    fun openXmlFile(inputStream: InputStream) = DomFileEditor(inputStream)
+}
+
+/**
+ * Wrapper for a file that can be edited as a dom document.
+ *
+ * @param inputStream the input stream to read the xml file from.
+ * @param onSave A callback that will be called when the editor is closed to save the file.
+ */
+class DomFileEditor internal constructor(
+    private val inputStream: InputStream,
+    private val onSave: ((String) -> Unit)? = null
```

Comment:
This parameter is never used and should be removed as it does not serve any specific purpose.

Reply:
I am not sure what you mean.
The parameter is used in the `close()` function. https://github.com/revanced/revanced-patcher/blob/6950efc52d442603370ac1020dc54c5491b13fc6/src/main/kotlin/app/revanced/patcher/Context.kt#L88-L101

Reply:
It is used here, but the constructor is not used anywhere in the project and is marked as internal.

Reply:
> It is used here, but the constructor is not used anywhere in the project and is marked as internal.

https://github.com/revanced/revanced-patcher/blob/d96cc5dd1ea90b1653b7a8577997c185218013b9/revanced-patcher/src/main/kotlin/app/revanced/patcher/Context.kt#L68

Reply:
Yes, but the second parameter is not used and the constructor is marked as internal, therefor the second parameter serves no purpose

Reply:
> Yes, but the second parameter is not used and the constructor is marked as internal, therefor the second parameter serves no purpose

The parameter is set by the secondary constructor that takes a `ResourceFile`. I don't understand what you want me to do.